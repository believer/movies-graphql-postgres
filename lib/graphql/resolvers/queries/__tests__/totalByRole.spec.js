const totalByRole = require('../totalByRole')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/totalByRole', () => {
  let db

  beforeEach(() => {
    db = {
      one: jest.fn().mockReturnValue({
        count: 1,
      }),
    }
  })

  it('should verify token', () => {
    totalByRole(null, { ranking: 10 }, { db, token: 'test' })

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should get how many movies a for a specific role', async () => {
    await totalByRole(
      null,
      {
        role: 'director',
      },
      { db }
    )

    expect(db.one.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await totalByRole(
      null,
      {
        role: 'director',
      },
      { db }
    )

    expect(response).toEqual(1)
  })

  it('should handle errors', async () => {
    db.one.mockImplementation(() => Promise.reject('nope'))

    try {
      await totalByRole(
        null,
        {
          role: 'director',
        },
        { db }
      )
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
