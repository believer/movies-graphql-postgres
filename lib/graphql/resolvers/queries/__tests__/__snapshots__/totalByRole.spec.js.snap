// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/totalByRole should get how many movies a for a specific role 1`] = `
"
      SELECT
        COUNT(distinct p.name)
      FROM 
        views AS v
        INNER JOIN movies AS m ON m.id = v.movie_id
        INNER JOIN cast_and_crew AS cac ON cac.movie_id = m.id
        INNER JOIN people AS p ON cac.person_id = p.id
      WHERE v.user_id = 2 AND cac.role_id = 1
    "
`;
